src/comparator.rs

```diff
@@ -1,30 +1,482 @@
+refint_input!(COMP2);
+
+macro_rules! dac_input {
+    ($COMP:ident, $channel:ty, $bits:expr) => {
+        impl NegativeInput<$COMP> for &$channel {
+            fn setup(&self, comp: &$COMP) {
+                comp.csr().modify(|_, w| unsafe { w.inmsel().bits($bits) })
+            }
+        }
+    };
+}
+
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP1, dac::Channel1<dac::Enabled>, 0b0100);
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP1, dac::Channel2<dac::Enabled>, 0b0101);
+
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP2, dac::Channel1<dac::Enabled>, 0b0100);
+#[cfg(any(feature = "stm32g071", feature = "stm32g081"))]
+dac_input!(COMP2, dac::Channel2<dac::Enabled>, 0b0101);
+
+pub struct Comparator<C> {
+    regs: C,
+}
+
+pub trait ComparatorExt<COMP> {
+    fn init<P: PositiveInput<COMP>, N: NegativeInput<COMP>>(
+        &mut self,
+        positive_input: P,
+        negative_input: N,
+        config: Config,
+    );
+    fn output(&self) -> bool;
+    fn enable(&self);
+    fn disable(&self);
+    fn output_pin<P: OutputPin<COMP>>(&self, pin: P);
+    fn listen(&self, edge: SignalEdge, exti: &EXTI);
+    fn unlisten(&self, exti: &EXTI);
+    fn is_pending(&self, edge: SignalEdge, exti: &EXTI) -> bool;
+    fn unpend(&self, exti: &EXTI);
+}
+
+macro_rules! comparator_ext {
+    ($COMP:ty, $Comparator:ty, $Event:expr) => {
+        impl ComparatorExt<$COMP> for $Comparator {
+            // TODO: Require init before any other functions?
+            fn init<P: PositiveInput<$COMP>, N: NegativeInput<$COMP>>(
+                &mut self,
+                positive_input: P,
+                negative_input: N,
+                config: Config,
+            ) {
+                positive_input.setup(&self.regs);
+                negative_input.setup(&self.regs);
+                self.regs.csr().modify(|_, w| unsafe {
+                    w.hyst()
+                        .bits(config.hysteresis as u8)
+                        .polarity()
+                        .bit(config.inverted)
+                        .pwrmode()
+                        .bits(config.power_mode as u8)
+                        .winout()
+                        .bit(config.output_xor)
+                });
+                // TODO: Delay for comp scaler bridge voltage stabilization?
+            }
+
+            fn output(&self) -> bool {
+                self.regs.csr().read().value().bit_is_set()
+            }
+
+            fn enable(&self) {
+                self.regs.csr().modify(|_, w| w.en().set_bit());
+                // TODO: Startup delay?
+            }
+
+            fn disable(&self) {
+                self.regs.csr().modify(|_, w| w.en().clear_bit());
+            }
+
+            fn output_pin<P: OutputPin<$COMP>>(&self, pin: P) {
+                pin.setup();
+            }
+
+            // TODO: Does the COMP need to be disabled when this is enabled?
```

Comment:
How ST hal is doing it?

Reply:
Looking at the RM and HAL again, I believe the comparator does need to be disabled. I'll add Enable/Disable typestate.